### PR TITLE
Added PUT ,PATCH and DELETE methods supports in XHR.

### PR DESCRIPTION
**Pull request by ddesai96**
XHR supports the PUT and DELETE method

### CODE REVIEW

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 1]
```diff
@@ -328,8 +356,14 @@ bool CurlNetworking::sendRequest(shared_ptr<CurlRequest> curlRequest, folly::dyn
   if(headers != nullptr) {
     setHeaders(curlRequest, headers);
   }
-  if(!(strcmp(curlRequest->method.c_str(), "POST"))) {
```
Comment (munezbn) [inline ReactSkia/sdk/CurlNetworking.cpp:331]: Add or statements here for similar methods.

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 2]
```diff
@@ -155,38 +155,66 @@ void CurlNetworking::processNetworkRequest(CURLM *curlMultiHandle) {
 
 }
 
-bool CurlNetworking::preparePostRequest(shared_ptr<CurlRequest> curlRequest, folly::dynamic data ) {
-  const char *dataPtr = NULL;
+size_t CurlNetworking::read_callback(void *ptr, size_t size, size_t nmemb, void *userdata)
+{
+  memcpy((char*)ptr,(char*)userdata,(size * nmemb));
```
Comment (munezbn) [inline ReactSkia/sdk/CurlNetworking.cpp]: How the buffer data is advanced ?

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 3]
```diff
@@ -9,7 +9,7 @@
 #include "ReactSkia/utils/RnsLog.h"
 #include "ReactSkia/utils/RnsUtils.h"
 #include "CurlNetworking.h"
-
+#include <string.h>
```
Comment (chatrapathi06) [inline ReactSkia/sdk/CurlNetworking.cpp]: is this headerfile is required

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 4]
```diff
@@ -155,38 +155,81 @@ void CurlNetworking::processNetworkRequest(CURLM *curlMultiHandle) {
...
+}
+
+
+bool CurlNetworking::prepareRequest(shared_ptr<CurlRequest> curlRequest, folly::dynamic data, string methodName ) {
   size_t dataSize = 0;
   bool status = false;
+  if(strcmp(methodName.c_str(),"DELETE")) {
+    if(data["string"].c_str()) {
+      dataSize = data["string"].getString().length();
+      dataPtr =(char *) malloc(dataSize);
```
Comment (munezbn) [inline ReactSkia/sdk/CurlNetworking.cpp]: When this will be freed ?

File: ReactSkia/sdk/CurlNetworking.h [hunk 5]
```diff
@@ -66,6 +66,9 @@ class CurlRequest {
   std::string URL;
   size_t timeout;
   std::string method;
+  unsigned int dataLength;
+  unsigned int requestDataOffset;
+  char *dataPtr;
```
Comment (munezbn) [inline ReactSkia/sdk/CurlNetworking.h]: can we prefix upload ?
uploadDataLength
uploadDataPtr
uploadBufferOffset

**Review by munezbn — APPROVED** [review]

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 6]
```diff
@@ -53,6 +53,12 @@ CurlRequest::CurlRequest(CURL *lhandle, std::string lURL, size_t ltimeout, std::
   curlResponse(make_shared<CurlResponse>())
   {}
 
+CurlRequest::~CurlRequest() {
+  if(uploadDataPtr!=NULL) {
+    free(uploadDataPtr);
```
Comment (shefayaparwin) [inline ReactSkia/sdk/CurlNetworking.cpp:58]: Can be set uploadDataptr to NULL after frreing it

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 7]
```diff
@@ -155,38 +161,85 @@ void CurlNetworking::processNetworkRequest(CURLM *curlMultiHandle) {
 
 }
 
-bool CurlNetworking::preparePostRequest(shared_ptr<CurlRequest> curlRequest, folly::dynamic data ) {
-  const char *dataPtr = NULL;
+size_t CurlNetworking::read_callback(void* ptr, size_t size, size_t nmemb, void* userdata) {
+
+  CurlRequest *curlRequest = (CurlRequest *)userdata;
```
Comment (shefayaparwin) [inline ReactSkia/sdk/CurlNetworking.cpp]: Can you verify the validatity of curlRequest handle before reading it

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 8]
```diff
@@ -155,38 +161,85 @@ void CurlNetworking::processNetworkRequest(CURLM *curlMultiHandle) {
...
+  size_t readSize = size * nmemb;
+
+  // Calculate the remaining data size to send
+  size_t remainingSize = curlRequest->uploadDataLength - curlRequest->uploadBufferOffset;
+
+  if(remainingSize == 0) { // no more data left to copy,stop the current transfer
```
Comment (shefayaparwin) [inline ReactSkia/sdk/CurlNetworking.cpp:173]: Is it == or <= 0

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 9]
```diff
@@ -155,38 +161,85 @@ void CurlNetworking::processNetworkRequest(CURLM *curlMultiHandle) {
 
 }
 
-bool CurlNetworking::preparePostRequest(shared_ptr<CurlRequest> curlRequest, folly::dynamic data ) {
-  const char *dataPtr = NULL;
+size_t CurlNetworking::read_callback(void* ptr, size_t size, size_t nmemb, void* userdata) {
```
Comment (shefayaparwin) [inline ReactSkia/sdk/CurlNetworking.cpp]: what is this nmemb means ? can it be explainable name

Comment (shefayaparwin) [inline ReactSkia/sdk/CurlNetworking.cpp]: Can you verify ptr pointer validity before proceeding

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 10]
```diff
@@ -155,38 +161,85 @@ void CurlNetworking::processNetworkRequest(CURLM *curlMultiHandle) {
...
-    RNS_LOG_NOT_IMPL;
-    return status;
-  } else if(data["base64"].c_str()) {
-    RNS_LOG_NOT_IMPL;
-    return status;
-  } else {
-    RNS_LOG_ERROR ("Unknown Data for Post Request\n") ;
-    return status;
+  if(strcmp(methodName.c_str(),"DELETE")) {
+    if(data["string"].c_str()) {
```
Comment (shefayaparwin) [inline ReactSkia/sdk/CurlNetworking.cpp]: Are you checking string is not empty ? then can you using ! empty() method itself for better clarifty

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 11]
```diff
@@ -155,38 +161,85 @@ void CurlNetworking::processNetworkRequest(CURLM *curlMultiHandle) {
...
+      dataSize = data["string"].getString().length();
+      curlRequest->uploadDataPtr =(char *) malloc(dataSize);
+      strcpy(curlRequest->uploadDataPtr,data["string"].c_str());
+    } else if(data["formData"].c_str()) {
+      RNS_LOG_NOT_IMPL;
```
Comment (shefayaparwin) [inline ReactSkia/sdk/CurlNetworking.cpp]: Can you add not implemented msh in detail of what is  not implemented instead here , it don't give clear info on what is not implemented

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 12]
```diff
@@ -155,38 +161,85 @@ void CurlNetworking::processNetworkRequest(CURLM *curlMultiHandle) {
...
-    return status;
-  } else if(data["base64"].c_str()) {
-    RNS_LOG_NOT_IMPL;
-    return status;
-  } else {
-    RNS_LOG_ERROR ("Unknown Data for Post Request\n") ;
-    return status;
+  if(strcmp(methodName.c_str(),"DELETE")) {
+    if(data["string"].c_str()) {
+      dataSize = data["string"].getString().length();
```
Comment (shefayaparwin) [inline ReactSkia/sdk/CurlNetworking.cpp:192]: does getString() is needed ?? data["string"] , itself a string right ?

**Review by shefayaparwin — COMMENTED** [review]
In  Genaral, suggest to  use below style for qtring compare to better readability. 
if (str1.compare(str2) != 0)

**Review by shefayaparwin — COMMENTED** [review]
In  Genaral, suggest to  use below style for qtring compare to better readability. 
if (str1.compare(str2) != 0)

File: ReactSkia/sdk/CurlNetworking.cpp (hunk 9 above)
Comment (shefayaparwin) [inline ReactSkia/sdk/CurlNetworking.cpp]: why the api name is read_callback instead readCallback ? any reason for deviating it from standard.

File: ReactSkia/sdk/CurlNetworking.cpp (hunk 7 above)
Comment (shefayaparwin) [inline ReactSkia/sdk/CurlNetworking.cpp]: it is good to use static_cast. can catch error better if any
static_cast<CurlRequest*>(userdata);

File: ReactSkia/sdk/CurlNetworking.cpp (hunk 8 above)
Comment (ddesai96) [reply ReactSkia/sdk/CurlNetworking.cpp:173]: yes it is ==

File: ReactSkia/sdk/CurlNetworking.cpp (hunk 12 above)
Comment (ddesai96) [reply ReactSkia/sdk/CurlNetworking.cpp:192]: yes ,getString() is needed.

File: ReactSkia/sdk/CurlNetworking.cpp [hunk 13]
```diff
@@ -155,38 +162,87 @@ void CurlNetworking::processNetworkRequest(CURLM *curlMultiHandle) {
 
 }
 
-bool CurlNetworking::preparePostRequest(shared_ptr<CurlRequest> curlRequest, folly::dynamic data ) {
-  const char *dataPtr = NULL;
+size_t CurlNetworking::readCallback(void* ptr, size_t size, size_t nitems, void* userdata) {
+
+  CurlRequest *curlRequest = (CurlRequest *)userdata;
+  if(!curlRequest) {
```
Comment (shefayaparwin) [inline ReactSkia/sdk/CurlNetworking.cpp:167]: REturn to be done if it is NULL

Comment (ddesai96) [reply ReactSkia/sdk/CurlNetworking.cpp:167]: the return type is size_t, so have to return zero

**Review by shefayaparwin — APPROVED** [review]

**Review by chatrapathi06 — APPROVED** [review]

**Pull request merged**